aiohomekit/controller/ble/client.py

```diff
@@ -164,72 +193,104 @@ async def ble_request(
             if next is False:
                 raise EncryptionError("Decryption failed")
-        if debug_enabled:
-            logger.debug("Read fragment: %s", next)
+        logger.debug("Read fragment: %s", next)
 
         data += decode_pdu_continuation(tid, next)
 
     return status, data
 
 
+def raise_for_pdu_status(client: BleakClient, pdu_status: PDUStatus) -> None:
+    """Raise on non-success PDU status."""
+    if pdu_status != PDUStatus.SUCCESS:
+        raise ValueError(
+            f"{client.address}: PDU status was not success: {pdu_status.description} ({pdu_status.value})"
+        )
+
+
+def _decode_pdu_tlv_value(
+    client: AIOHomeKitBleakClient, pdu_status: PDUStatus, data: bytes
+) -> bytes:
+    """Decode the TLV value from the PDU."""
+    raise_for_pdu_status(client, pdu_status)
+    decoded = dict(TLV.decode_bytes(data))
+    return decoded[AdditionalParameterTypes.Value.value]
+
+
 async def char_write(
     client: BleakClient,
     encryption_key: EncryptionKey | None,
     decryption_key: DecryptionKey | None,
     handle: BleakGATTCharacteristic,
     iid: int,
     body: bytes,
-):
+) -> bytes:
+    """Execute a CHAR_WRITE request."""
     body = BleRequest(expect_response=1, value=body).encode()
     pdu_status, data = await ble_request(
         client, encryption_key, decryption_key, OpCode.CHAR_WRITE, handle, iid, body
     )
-    if pdu_status != PDUStatus.SUCCESS:
-        raise ValueError(
-            f"{client.address}: PDU status was not success: {pdu_status.description} ({pdu_status.value})"
-        )
-    decoded = dict(TLV.decode_bytes(data))
-    return decoded[AdditionalParameterTypes.Value.value]
+    return _decode_pdu_tlv_value(client, pdu_status, data)
+
+
+async def _pairing_char_write(
+    client: AIOHomeKitBleakClient,
+    handle: BleakGATTCharacteristic,
+    iid: int,
+    request: list[tuple[TLV, bytes]],
+) -> dict[int, bytes]:
+    """Read or write a characteristic value."""
+    buffer = bytearray()
+    next_write = TLV.encode_list(request)
+
+    for _ in range(MAX_REASSEMBLY):
+        data = await char_write(client, None, None, handle, iid, next_write)
+        decoded = dict(TLV.decode_bytearray(data))
+        if TLV.kTLVType_FragmentLast in decoded:
+            logger.debug("%s: Reassembling final fragment", client.address)
+            buffer.extend(decoded[TLV.kTLVType_FragmentLast])
+            return dict(TLV.decode_bytes(buffer))
+        elif TLV.kTLVType_FragmentData in decoded:
+            logger.debug("%s: Reassembling fragment", client.address)
+            # There is more data, acknowledge the fragment
+            # and keep reading
+            buffer.extend(decoded[TLV.kTLVType_FragmentData])
+            # Acknowledge the fragment
+            # We must construct this manually since TLV.encode_bytes
+            # current does not know how to encode a 0 length
+            next_write = bytes([TLV.kTLVType_FragmentData, 0])
```

Comment:
This was the key thing to figure out since the TLV encoder can't encode 0 length